substrate-runtime/src/finality.rs

```diff
@@ -0,0 +1,147 @@
+            Ok(())
+        }
+
+        pub fn issue_signature(origin, which: u32, sig: Signature) -> Result {
+            let who = ensure_signed(origin)?;
+            ensure!(<Validators<T>>::exists(who.clone()), "Called by non validator");
+            let max = <MaxIndex<T>>::take();
+            ensure!(which < max, "Invalid block number");
+            let public = <Validators<T>>::get(who);
+
+            let vault_hash = <VaultChain<T>>::get(which);
+            ensure!(verify(vault_hash, &sig, &public), "Invalid Signature");
+            Self::deposit_event(Event::Signed(which, vault_hash, sig, public));
+
+            ensure!(<VoteCount<T>>::exists(which), "Can't vote on proof for this endpoint");
+            let mut status : (u32, u32) = <VoteCount<T>>::get(which);
+            status.0 += 1_u32;
+            <VoteCount<T>>::insert(which, status);
+            Self::check_finality(which, &mut status, vault_hash);
+
+            Ok(())
+        }
+
+        pub fn issue_rebuke(origin, which: u32) -> Result {
+            let who = ensure_signed(origin)?;
+            ensure!(<Validators<T>>::exists(who.clone()), "Called by non validator");
+            let max = <MaxIndex<T>>::take();
+            ensure!(which < max, "Invalid block number");
+            let public = <Validators<T>>::get(who);
+
+            let vault_hash = <VaultChain<T>>::get(which);
+            Self::deposit_event(Event::Rebuked(which, vault_hash, public));
+
+            ensure!(<VoteCount<T>>::exists(which), "Can't vote on proof for this endpoint");
+            let mut status : (u32, u32) = <VoteCount<T>>::get(which);
+            status.1 += 1_u32;
+            <VoteCount<T>>::insert(which, status);
+            Self::check_finality(which, &mut status, vault_hash);
+
+            Ok(())
+        }
+    }
+}
+
+impl<T: Trait> Module<T> {
+    fn check_finality(which: u32, status: &mut (u32, u32), vault_hash: [u8; 32]) {
+        let number_of_validators = <NumValidators<T>>::take();
+        if status.0 + status.1 == number_of_validators {
+            // TODO - Abstract as adjustable constants [it is set as 2/3 approval needed
+            // right now]
+            if 3 * status.0 > 2 * number_of_validators {
+                Self::deposit_event(Event::Finalized(which, vault_hash));
+            } else {
+                Self::deposit_event(Event::Failed(which, vault_hash));
+            }
+        }
+    }
+
+    pub fn add_hash(vault_hash: [u8; 32]) {
+        let index = <MaxIndex<T>>::take();
+        <VaultChain<T>>::insert(index, vault_hash);
+        <MaxIndex<T>>::put(index + 1);
+    }
+}
+
+// TODO - Replace with a hash type from the merkle system in stark.
+type HashContainer = [u8; 32]; // We need a type alias to avoid brackets in the macro.
+
+decl_event! {
+    pub enum Event
+    {
+        ProofRecorded(u32, u32, RecordedProof),
+        Signed(u32, HashContainer, Signature, PublicKey),
+        Rebuked(u32, HashContainer, PublicKey),
```

Comment:
Should this be Revoked?

Reply:
It's emitted by a single validator so the hash hasn't been revoked, just one person doesn't approve. Plus revoked implies that something was had but in this case it should only be called on something which has no judgement yet.